File: src/algebra2_logarithm_properties_medium.py
"""
algebra2 - logarithm_properties (medium)
Generated: 2026-02-22T04:31:04.009558
"""

from problem_utils import *

def generate():
    problem_type = randint(1, 6)

    if problem_type == 1:
        # Expand: log_b(m*n) or log_b(x*y)
        base = choice([2, 3, 5, 7, 10])
        coeff1 = randint(2, 15)
        coeff2 = randint(2, 15)

        if randint(0, 1) == 0:
            question = f"\\log_{{{base}}}({coeff1} \\cdot {coeff2})"
            ans = f"\\log_{{{base}}}({coeff1}) + \\log_{{{base}}}({coeff2})"
            solution_text = steps(
                f"Apply the product rule: $\\log_b(mn) = \\log_b(m) + \\log_b(n)$",
                f"${question} = {ans}$"
            )
        else:
            vars_pair = choice([("x", "y"), ("a", "b"), ("m", "n")])
            u, v = vars_pair
            question = f"\\log_{{{base}}}({u}{v})"
            ans = f"\\log_{{{base}}}({u}) + \\log_{{{base}}}({v})"
            solution_text = steps(
                f"Apply the product rule: $\\log_b(mn) = \\log_b(m) + \\log_b(n)$",
                f"${question} = {ans}$"
            )

        return problem(
            question=f"Expand using logarithm properties: ${question}$",
            answer=ans,
            difficulty=(1300, 1400),
            topic="algebra2/logarithm_properties",
            solution=solution_text,
            answer_type="expression"
        )

    elif problem_type == 2:
        # Expand quotient rule
        base = choice([2, 3, 5, 7, 10])

        if randint(0, 1) == 0:
            num = randint(6, 60)
            denom = randint(2, 12)
            question = f"\\log_{{{base}}}\\left(\\frac{{{num}}}{{{denom}}}\\right)"
            ans = f"\\log_{{{base}}}({num}) - \\log_{{{base}}}({denom})"
            solution_text = steps(
                f"Apply the quotient rule: $\\log_b\\left(\\frac{{m}}{{n}}\\right) = \\log_b(m) - \\log_b(n)$",
                f"${question} = {ans}$"
            )
        else:
            vars_pair = choice([("x", "y"), ("a", "b"), ("p", "q")])
            u, v = vars_pair
            question = f"\\log_{{{base}}}\\left(\\frac{{{u}}}{{{v}}}\\right)"
            ans = f"\\log_{{{base}}}({u}) - \\log_{{{base}}}({v})"
            solution_text = steps(
                f"Apply the quotient rule: $\\log_b\\left(\\frac{{m}}{{n}}\\right) = \\log_b(m) - \\log_b(n)$",
                f"${question} = {ans}$"
            )

        return problem(
            question=f"Expand using logarithm properties: ${question}$",
            answer=ans,
            difficulty=(1300, 1400),
            topic="algebra2/logarithm_properties",
            solution=solution_text,
            answer_type="expression"
        )

    elif problem_type == 3:
        # Expand power rule
        base = choice([2, 3, 5, 7, 10])
        power = randint(2, 6)
        var = choice(["x", "a", "m", "n"])

        if randint(0, 1) == 0:
            value = randint(2, 15)
            question = f"\\log_{{{base}}}({value}^{{{power}}})"
            ans = f"{power} \\log_{{{base}}}({value})"
        else:
            question = f"\\log_{{{base}}}({var}^{{{power}}})"
            ans = f"{power} \\log_{{{base}}}({var})"

        solution_text = steps(
            f"Apply the power rule: $\\log_b(m^n) = n \\log_b(m)$",
            f"${question} = {ans}$"
        )

        return problem(
            question=f"Expand using logarithm properties: ${question}$",
            answer=ans,
            difficulty=(1300, 1400),
            topic="algebra2/logarithm_properties",
            solution=solution_text,
            answer_type="expression"
        )

    elif problem_type == 4:
        # Combined expansion: log(x^a * y^b / z^c) or log(x^a * y)
        base = choice([2, 3, 5, 7, 10])
        p1 = randint(2, 5)
        p2 = randint(2, 5)
        p3 = randint(2, 4)

        variant = randint(0, 2)
        if variant == 0:
            question = f"\\log_{{{base}}}\\left(\\frac{{x^{{{p1}}} y^{{{p2}}}}}{{z^{{{p3}}}}}\\right)"
            ans = f"{p1}\\log_{{{base}}}(x) + {p2}\\log_{{{base}}}(y) - {p3}\\log_{{{base}}}(z)"
            solution_text = steps(
                f"Apply quotient rule: split numerator from denominator",
                f"Apply product rule to numerator: $\\log_{{{base}}}(x^{{{p1}}} y^{{{p2}}}) = \\log_{{{base}}}(x^{{{p1}}}) + \\log_{{{base}}}(y^{{{p2}}})$",
                f"Apply power rule: $= {ans}$"
            )
        elif variant == 1:
            question = f"\\log_{{{base}}}(x^{{{p1}}} y^{{{p2}}})"
            ans = f"{p1}\\log_{{{base}}}(x) + {p2}\\log_{{{base}}}(y)"
            solution_text = steps(
                f"Apply product rule: $\\log_b(AB) = \\log_b(A) + \\log_b(B)$",
                f"Apply power rule to each: $= {ans}$"
            )
        else:
            question = f"\\log_{{{base}}}\\left(\\frac{{x^{{{p1}}}}}{{y^{{{p2}}} z^{{{p3}}}}}\\right)"
            ans = f"{p1}\\log_{{{base}}}(x) - {p2}\\log_{{{base}}}(y) - {p3}\\log_{{{base}}}(z)"
            solution_text = steps(
                f"Apply quotient rule then power rule",
                f"$= {ans}$"
            )

        return problem(
            question=f"Expand using logarithm properties: ${question}$",
            answer=ans,
            difficulty=(1400, 1550),
            topic="algebra2/logarithm_properties",
            solution=solution_text,
            answer_type="expression"
        )

    elif problem_type == 5:
        # Condense: a*log(x) + b*log(y) -> log(x^a * y^b)
        base = choice([2, 3, 5, 7, 10])
        c1 = randint(2, 5)
        c2 = randint(2, 5)
        var1 = choice(["x", "a", "m"])
        var2 = choice(["y", "b", "n"])
        while var2 == var1:
            var2 = choice(["y", "b", "n"])

        question = f"{c1}\\log_{{{base}}}({var1}) + {c2}\\log_{{{base}}}({var2})"
        ans = f"\\log_{{{base}}}({var1}^{{{c1}}} {var2}^{{{c2}}})"
        solution_text = steps(
            f"Apply power rule: $n\\log_b(m) = \\log_b(m^n)$",
            f"$= \\log_{{{base}}}({var1}^{{{c1}}}) + \\log_{{{base}}}({var2}^{{{c2}}})$",
            f"Apply product rule: $= {ans}$"
        )

        return problem(
            question=f"Write as a single logarithm: ${question}$",
            answer=ans,
            difficulty=(1350, 1500),
            topic="algebra2/logarithm_properties",
            solution=solution_text,
            answer_type="expression"
        )

    else:
        # Condense: a*log(x) - log(y)
        base = choice([2, 3, 5, 7, 10])
        coeff = randint(2, 5)
        var1 = choice(["x", "a", "m"])
        var2 = choice(["y", "b", "n"])
        while var2 == var1:
            var2 = choice(["y", "b", "n"])

        question = f"{coeff}\\log_{{{base}}}({var1}) - \\log_{{{base}}}({var2})"
        ans = f"\\log_{{{base}}}\\left(\\frac{{{var1}^{{{coeff}}}}}{{{var2}}}\\right)"
        solution_text = steps(
            f"Apply power rule: ${coeff}\\log_{{{base}}}({var1}) = \\log_{{{base}}}({var1}^{{{coeff}}})$",
            f"Apply quotient rule: $\\log_b(m) - \\log_b(n) = \\log_b(m/n)$",
            f"$= {ans}$"
        )

        return problem(
            question=f"Write as a single logarithm: ${question}$",
            answer=ans,
            difficulty=(1350, 1500),
            topic="algebra2/logarithm_properties",
            solution=solution_text,
            answer_type="expression"
        )

emit(generate())

File: src/algebra2_logarithm_properties_medium.jl
# algebra2 - logarithm_properties (medium)
# Generated: 2026-03-08T20:20:24.800174

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x begin
    set_topic!("algebra2/logarithm_properties")
    
    problem_type = rand(1:5)
    
    if problem_type == 1
        # Expanding logarithms using product/quotient/power rules
        # ELO: 1200-1400
        base = choice([2, 3, 5, 10])
        a = randint(2, 12)
        b = randint(2, 12)
        n = randint(2, 5)
        
        use_product = rand(Bool)
        use_power = rand(Bool)
        
        if use_product && use_power
            ans_str = "$(n)\\log_{$(base)}($(a)) + \\log_{$(base)}($(b))"
            question_str = "\\log_{$(base)}($(a)^{$(n)} \\cdot $(b))"
        elseif use_product
            ans_str = "\\log_{$(base)}($(a)) + \\log_{$(base)}($(b))"
            question_str = "\\log_{$(base)}($(a) \\cdot $(b))"
        else
            ans_str = "\\log_{$(base)}($(a)) - \\log_{$(base)}($(b))"
            question_str = "\\log_{$(base)}\\left(\\frac{$(a)}{$(b)}\\right)"
        end
        
        problem(
            question="Expand using logarithm properties: \$$(question_str)\$",
            answer=ans_str,
            difficulty=(1200, 1400),
            answer_type="expression",
            solution=steps(
                "Use logarithm properties:",
                use_product ? "Product rule: \\(\\log_b(xy) = \\log_b(x) + \\log_b(y)\\)" : "Quotient rule: \\(\\log_b(x/y) = \\log_b(x) - \\log_b(y)\\)",
                use_power ? "Power rule: \\(\\log_b(x^n) = n\\log_b(x)\\)" : "",
                sol("Answer", ans_str)
            ),
            time=90
        )
        
    elseif problem_type == 2
        # Condensing logarithms
        # ELO: 1300-1500
        base = choice([2, 3, 5, 10])
        a = randint(2, 8)
        b = randint(2, 8)
        c = randint(2, 5)
        d = randint(2, 5)
        
        operation = rand(1:3)
        if operation == 1
            # Sum to product
            question_str = "$(c)\\log_{$(base)}($(a)) + $(d)\\log_{$(base)}($(b))"
            ans_str = "\\log_{$(base)}($(a)^{$(c)} \\cdot $(b)^{$(d)})"
        elseif operation == 2
            # Difference to quotient
            question_str = "$(c)\\log_{$(base)}($(a)) - $(d)\\log_{$(base)}($(b))"
            ans_str = "\\log_{$(base)}\\left(\\frac{$(a)^{$(c)}}{$(b)^{$(d)}}\\right)"
        else
            # Mixed
            e = randint(2, 8)
            question_str = "\\log_{$(base)}($(a)) + $(c)\\log_{$(base)}($(b)) - \\log_{$(base)}($(e))"
            ans_str = "\\log_{$(base)}\\left(\\frac{$(a) \\cdot $(b)^{$(c)}}{$(e)}\\right)"
        end
        
        problem(
            question="Condense into a single logarithm: \$$(question_str)\$",
            answer=ans_str,
            difficulty=(1300, 1500),
            answer_type="expression",
            solution=steps(
                "Apply logarithm properties in reverse:",
                "Power rule: \\(n\\log_b(x) = \\log_b(x^n)\\)",
                operation == 1 ? "Product rule: \\(\\log_b(x) + \\log_b(y) = \\log_b(xy)\\)" : "Quotient rule: \\(\\log_b(x) - \\log_b(y) = \\log_b(x/y)\\)",
                sol("Answer", ans_str)
            ),
            time=120
        )
        
    elseif problem_type == 3
        # Solving logarithmic equations (single log)
        # ELO: 1400-1600
        base = choice([2, 3, 4, 5, 10])
        ans = randint(2, 50)
        value = randint(2, 6)
        
        # log_base(x) = value, so x = base^value
        ans = base^value
        
        problem(
            question="Solve for \\(x\\): \$\\log_{$(base)}(x) = $(value)\$",
            answer=ans,
            difficulty=(1400, 1600),
            solution=steps(
                sol("Given", "\\log_{$(base)}(x) = $(value)"),
                "Convert to exponential form: \\(x = $(base)^{$(value)}\\)",
                sol("Answer", "x = $(ans)")
            ),
            time=90
        )
        
    elseif problem_type == 4
        # Solving logarithmic equations (combining logs)
        # ELO: 1500-1700
        base = choice([2, 3, 5, 10])
        ans = randint(3, 20)
        other = randint(2, 8)
        
        while other == ans
            other = randint(2, 8)
        end
        
        # log_base(x) + log_base(other) = log_base(x * other)
        product = ans * other
        result_power = randint(1, 3)
        
        # Find what power of base equals product
        result = product
        
        problem(
            question="Solve for \\(x\\): \$\\log_{$(base)}(x) + \\log_{$(base)}($(other)) = \\log_{$(base)}($(product))\$",
            answer=ans,
            difficulty=(1500, 1700),
            solution=steps(
                sol("Given", "\\log_{$(base)}(x) + \\log_{$(base)}($(other)) = \\log_{$(base)}($(product))"),
                "Use product rule: \\(\\log_{$(base)}(x \\cdot $(other)) = \\log_{$(base)}($(product))\\)",
                "Since the logarithms are equal: \\(x \\cdot $(other) = $(product)\\)",
                "Divide both sides by $(other): \\(x = $(ans)\\)",
                sol("Answer", "x = $(ans)")
            ),
            time=150
        )
        
    else
        # Change of base formula application
        # ELO: 1400-1700
        old_base = choice([2, 3, 5, 7])
        new_base = choice([10, "e"])
        arg = randint(5, 30)
        
        if new_base == 10
            ans_str = "\\frac{\\log($(arg))}{\\log($(old_base))}"
        else
            ans_str = "\\frac{\\ln($(arg))}{\\ln($(old_base))}"
        end
        
        problem(
            question="Rewrite using the change of base formula with base $(new_base == 10 ? "10" : "e"): \$\\log_{$(old_base)}($(arg))\$",
            answer=ans_str,
            difficulty=(1400, 1700),
            answer_type="expression",
            solution=steps(
                sol("Given", "\\log_{$(old_base)}($(arg))"),
                "Change of base formula: \\(\\log_b(x) = \\frac{\\log_c(x)}{\\log_c(b)}\\)",
                new_base == 10 ? "Using base 10:" : "Using base e (natural log):",
                sol("Answer", ans_str)
            ),
            time=120
        )
    end
end